### PR TITLE
LibGfx+PixelPaint: Constrain selection to the bounds of the image

### PR DESCRIPTION
**Pull request by tcl3**
This PR includes several changes to selection behavior to bring it more in line with other image editing programs. It also ensures consistent behavior is across all selection tools.

The following changes have been made:

* The lasso tool can now make a selection outside of the currently active layer. This makes its behavior consistent with other selection tools.
* The lasso tool now has a crosshair cursor.
* Selections are now constrained to within the bounds of the image. The preview lines for the lasso and polygonal select tools can still be drawn outside the bounds of the image, while the rectangle tool preview is not shown outside the bounds of the image. This matches the behavior of both GIMP and Photoshop.

Video:

https://user-images.githubusercontent.com/2817754/228615024-980fd0a2-e734-48d6-8c7b-e7ada8382054.mp4

### CODE REVIEW

**Comment by GMTA** [comment]
I'm wondering why we even allow selecting outside of the image/content bounds? It's fine to move the tool there, but the actual selected pixels should stay within the content bounds in my opinion.

This solves your excessive memory usage issue as well as the differences to other paint programs that we have. Ideally, drawing a lasso selection is a simple path that we track, and upon release of the mouse we select all pixels within the content bounds.

**Comment by tcl3** [comment]
> I'm wondering why we even allow selecting outside of the image/content bounds? It's fine to move the tool there, but the actual selected pixels should stay within the content bounds in my opinion.

I agree. I had intended to make a PR to change this behavior I think the best thing to do is to open a separate PR to make this change and rebase this one onto it. This would also fix the issue we're currently seeing in #17593.

**Comment by tcl3** [comment]
> I think the best thing to do is to open a separate PR to make this change and rebase this one onto it. This would also fix the issue we're currently seeing in #17593.

I've brought the change that constrains the selection to the bounds of the image into this PR. It was more difficult than expected to separate the changes into two PRs.

I have updated the name and description of this PR to fully describe the changes made.

**Comment by tcl3** [comment]
> I think the best thing to do is to open a separate PR to make this change and rebase this one onto it. This would also fix the issue we're currently seeing in #17593.

I've brought the change that constrains the selection to the bounds of the image into this PR. It was more difficult than expected to separate the changes into two PRs.

I have updated the name and description of this PR to fully describe the changes made.

**Comment by AtkinsSJ** [comment]
Snapping the polygonal selection vertices inside the image rect feels wrong to me. I would expect it to behave that all the pixels that were inside the selection shape become selected instead, like the lasso tool.

**Comment by tcl3** [comment]
> Snapping the polygonal selection vertices inside the image rect feels wrong to me. I would expect it to behave that all the pixels that were inside the selection shape become selected instead, like the lasso tool.

You're absolutely right, that's a much better way for it to behave.

The lasso tool behaved the same way, but the problem was harder to spot because the line segments it creates are smaller and I didn't draw anything particularly concave in my video above.

I've had another crack at it. This time I've made the polygonal select and lasso tools behave as you suggested.

Here's a video showing how it behaves now:

https://user-images.githubusercontent.com/2817754/232190720-98840068-743e-4193-84e6-50ff1108d419.mp4

---

The downside to this approach is that I am now creating a bitmap that encloses the entire polygon (even the parts that go outside the canvas). I'm sure there's a way to achieve the same effect, while limiting the size of the polygon bitmap to the size of the image. I've added some FIXMEs that say that. As the performance is no worse than before, I think its fine for now.

File: Userland/Libraries/LibGfx/Rect.h [hunk 1]
```diff
@@ -604,19 +604,11 @@ class Rect {
...
-        if (y() < constrain_rect.y())
-            move_y = y() - constrain_rect.y();
-        auto rect = *this;
-        if (move_x != 0 || move_y != 0)
-            rect.translate_by(move_x, move_y);
-        return rect;
+        Point<T> top_left_point { top_left() };
+        Point<T> bottom_right_point { x() + width(), y() + height() };
+        top_left_point.constrain(constrain_rect);
+        bottom_right_point.constrain(constrain_rect);
```
Comment (GMTA) [inline Userland/Libraries/LibGfx/Rect.h:610]: If the top left point falls outside of `constrain_rect`, but the bottom right falls inside of it, and the rect's size is small enough that it fits entirely inside of `constrain_rect`, we are now resizing it which seems incorrect.

I think the logic should be:
1) Try to move rect inside the constrain rect, aligning the top left points where necessary
2) Resize the rect to be smaller if it doesn't fit inside the constrain rect

Maybe we should have some tests for this as well :^)

Comment (tcl3) [reply Userland/Libraries/LibGfx/Rect.h]: That makes sense. I think I've misunderstood the intent of this function. For what I'm doing, I don't want the rect to move if it's partially outside the input rect's bounds, so I've introduced a new function `Rect::clipped_to()`, which does the same thing as I've implemented here. I'm open to suggestions for a better name though.

Comment (GMTA) [reply Userland/Libraries/LibGfx/Rect.h]: I feel as though this new `::clipped_to` is very similar to what `::intersect` does?

Comment (tcl3) [reply Userland/Libraries/LibGfx/Rect.h]: `::intersected` works fine here. I've removed it.

**Pull request merged**